### DISCUSSION TITLE
ICU4X 0.3 RFD

### DISCUSSION THREAD

**Discussion by zbraniecki** [unicode-org/icu4x#634] [General]
Hi all,

Let's try to use this new tool :)

I've been reviewing all the work we've done since October on 0.2.

There's a lot, but we also missed a bunch of items that we wanted to focus on like WASM, FFI, Static, Async.

Additionally, with major features like Timezones and Components landing right by the release date it becomes unreasonable for me to work on performance evaluation and optimizations since they require feature stability (de-hardcoding DTF Provider lifetimes is a large task and bitrots very easily).
The result is that while 0.2 is feature-ready for Mozilla's internal testing - we can plug Plurals, Locale, DTF into SpiderMonkey/Gecko and compare it to current model and collect experience on API integration, etc., it is not optimal for performance/memory as we know about a number of low hanging fruits that will hopefully significantly impact perf characteristics for our testing.

My plan was to drive the release of 0.2 and then work with the group to hand over planning of the next major release somewhere by the EOQ or later which would focus on further completing the feature set for 1.0, but seeing where we are I'd like to suggest taking a pause on features and carving out a new release - 0.3 - relatively soon after 0.2.

The release would be much lighter on new features and focus on extending the "cleanup" tik-tok cycle phase into a release.
There would still be room for feature work in three areas but the rest would be just stability, performance, and cleanups.

Release Driver: @zbraniecki
Target: May 2021
[Milestone](https://github.com/unicode-org/icu4x/milestone/12)

## Proposed Scope of 0.3:

* [ ] DateTimeFormat
  * [ ] DataProvider Data lifetimes
  * [ ] Pattern lifetimes
  * [ ] Pattern Parser zero-allocation and use of icu_pattern for combination patterns
  * [ ] Data struct splitting (#257)
  * [ ] Address perf/mem regression vs 0.1 (#603)
* [ ] FFI
  * [ ] ECMA-402 Traits updates matching feature scope we have
* [ ] Locale
  * [ ] Data canonicalization (#218)
* [ ] General
  * [ ] Performance analysis and optimizations
  * [ ] Memory analysis and optimizations
  * [ ] Tech Debt reduction, cleanups, docs etc.

### Happening during 0.3, but not necessarily ready by release (so, target 0.4):

* [ ] Data Provider
  * [ ] Static
* [ ] FFI
  * [ ] WASM exploration (#207)
  * [ ] C/C++ bindings model (aligned between generic ICU4X FFI and Mozilla Unified Architecture)
* [ ] Unicode Properties
  * [ ] L3b, preparing for Segmentation and Regexp needs
* [ ] Segmenter

I'm happy to carry-over the driving of the release since I see it almost as "0.2.5", but also if someone else wants to take over, feel free. I think the next big one is 0.4, so my preference would be finding a driver for that one :)

WDYT?

**Comment by dminor** [comment]
That sounds good to me. I'd like to get https://github.com/unicode-org/icu4x/issues/218 in to a 0.3 release, since ideally it would have landed with the likely subtags work in 0.2.

**Comment by sffc** [comment]
This sounds good.  I'd like to consider #257 in scope for this release as well.

I suggest we target 0.3 for the end of May, and then the feature-driven 0.4 release for end of September.

**Reply by zbraniecki** [reply]
I'd definitely consider #257, #218 and #603 for 0.3.

**Comment by zbraniecki** [comment]
Thanks for the feedback so far! I feel encouraged to start planning for it. I renamed the old `0.3` milestone to `0.4` and created a new one called `0.3`. Please, add all your work planned for 0.3 into it with the focus on cleanups, performance and minor completeness additions.

If you work on major features and will have them in time for 0.3, I'm happy to include them, but the feature release is now 0.4 and I'd love to suggest rotating to a new release driver for it! :)
